http_server/request_testing/request_listener.go

```diff
@@ -0,0 +1,54 @@
+package requesttesting
+
+import (
+	"net"
+	"errors"
+	"log"
+)
+
+type mockListener struct {
+	connChannel chan net.Conn
+	serverEndpoint net.Conn
+}
+
+// For each request, we create an HTTP listener that passes the request provided
+// as a function parameter to the server.
+func newMockListener(endpoint  net.Conn) mockListener {
+
+	channel := make(chan net.Conn, 1)
+	channel <- endpoint
+	listener := mockListener{
+		connChannel: channel,
+		serverEndpoint: endpoint,
+	}
+	
+	return listener
+}
+
+// Passes an endpoint to the server. Weirdly enough, when I make one request,
+// Accept() and hence Close() are being called twice, but the handler only once.
```

Comment:
Do you have any idea why Accept() and Close() are called twice? This happens for every test and seems wasteful to me/

Reply:
So the `http.Server` will call `listener.Accept` continously. It's up to Accept to block until there is a new connection or there is an error. When a new connection appears the server will send it of in a new go routine to handle the connection and the server will then continue to wait for more new connections by calling Accept again.

Close is called twice since you call `defer listener.Close()` and `defer server.Shutdown()`. The call to `server.Shutdown()` calls listener.Close() internally.